### PR TITLE
Support non-transactional distibuted_exec

### LINKED ISSUE #1660: Function distributed_exec always transactional
When using `distributed_exec` for executing statements that cannot be executed inside a transaction, an error is given. For example:

```
=> SELECT * FROM distributed_exec($$ CREATE Database test $$);
ERROR:  [dn2]: CREATE DATABASE cannot run inside a transaction block
```

This behavior is not possible to disable.

Suggested improvement is to add a parameter to `distributed_exec` allowing the distributed execution to run without a transactional context.

### PR DESCRIPTION
**Pull request by mkindahl**
Before this commit, executions of `distributed_exec` was always
transactional and this could not be disabled. This cause problems when
executing statements that cannot be executed inside a transaction, such
as `CREATE DATABASE`.

With this commit, we introduce a parameter `transactional` to
`distributed_exec` that allow non-transactional executions on data
nodes. The parameter defaults to `TRUE`.

It also changes `distributed_exec` to be a procedure, which better
matches it's usage (it returns `VOID`).

Closes #1660

### CODE REVIEW

File: tsl/test/expected/dist_hypertable-11.out [hunk 1]
```diff
@@ -13,9 +13,9 @@ SELECT (add_data_node (name, host => 'localhost', DATABASE => name)).*
 FROM (VALUES (:'DATA_NODE_1'), (:'DATA_NODE_2'), (:'DATA_NODE_3')) v (name);
      node_name     |   host    | port  |     database      | node_created | database_created | extension_created 
 -------------------+-----------+-------+-------------------+--------------+------------------+-------------------
- dist_hypertable_1 | localhost | 55432 | dist_hypertable_1 | t            | t                | t
- dist_hypertable_2 | localhost | 55432 | dist_hypertable_2 | t            | t                | t
- dist_hypertable_3 | localhost | 55432 | dist_hypertable_3 | t            | t                | t
+ dist_hypertable_1 | localhost | 15432 | dist_hypertable_1 | t            | t                | t
```
Comment (erimatnor) [inline tsl/test/expected/dist_hypertable-11.out:16]: This change seems unrelated to the PR. I think the old port number is cached in your build directory and you need to recreate it.

File: tsl/test/expected/dist_commands.out [hunk 2]
```diff
@@ -344,3 +324,25 @@ ERROR:  data node "myserver" is not a TimescaleDB server
...
+                            database => 'dist_commands_1');
+    node_name    |   host    | port  |    database     | node_created | database_created | extension_created 
+-----------------+-----------+-------+-----------------+--------------+------------------+-------------------
+ dist_commands_1 | localhost | 55432 | dist_commands_1 | t            | t                | t
+(1 row)
+
+\set ON_ERROR_STOP 0
+CALL distributed_exec('CREATE DATABASE dist_commands_magic',
+       node_list => '{dist_commands_1}');
+ERROR:  [dist_commands_1]: CREATE DATABASE cannot run inside a transaction block
```
Comment (erimatnor) [inline tsl/test/expected/dist_commands.out:344]: This error doesn't make much sense for this case since there was no explicit transaction block. I understand why it happens, because we start a transaction block remotely, but it is still confusing.

Besides, the function _should_ be tested locally with both a transaction block and without it, using transactional=true and false.

Comment (mkindahl) [reply tsl/test/expected/dist_commands.out:374]: Added tests to test both inside and outside a transaction. Executing a non-transactional `distributed_exec` inside a transaction does not make sense, so added an error for that.

File: tsl/src/remote/dist_commands.c [hunk 3]
```diff
@@ -427,18 +429,36 @@ ts_dist_cmd_close_prepared_command(PreparedDistCmd *command)
 	list_free_deep(command);
 }
 
+static const char *
+get_procedure_name(FunctionCallInfo fcinfo)
```
Comment (svenklemm) [inline tsl/src/remote/dist_commands.c:433]: this is copy of postgres get_func_name, you should use that instead

Comment (mkindahl) [reply tsl/src/remote/dist_commands.c:433]: Ah, didn't see it. Will remove.

Comment (mkindahl) [reply tsl/src/remote/dist_commands.c]: Removed

**Review by pmwkaa — APPROVED** [review]
Looks good to me.

Not sure, do we have tests which runs distributed_exec() inside active transaction but with transactional => false?

Some thoughts: another approach to this functionality is that we could track if the current session runs inside transaction already. If that is the case distributed_exec() could be transaсtional based on that as well.

**Comment by mkindahl** [comment]
> Looks good to me.
> 
> Not sure, do we have tests which runs distributed_exec() inside active transaction but with transactional => false?

Since this commit adds the ability to run `distributed_exec` outside a transaction, we do not have any before this, but one is added in the commit.

> Some thoughts: another approach to this functionality is that we could track if the current session runs inside transaction already. If that is the case distributed_exec() could be transaсtional based on that as well.

Yes, that is a possibility, but for this step I think it is better that we are explicit.

**Review by erimatnor — APPROVED** [review]
